### PR TITLE
[Parser] Do not eagerly lex keywords

### PR DESCRIPTION
**Pull request by tlively**
Lex them on demand instead to avoid wasted work.

### CODE REVIEW

**Comment by tlively** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6541?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#6543** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6543?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6542** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6542?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6541** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6541?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#6540** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6540?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @tlively and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/parser/lexer.cpp [hunk 1]
```diff
@@ -1069,6 +1069,57 @@ bool Lexer::takeRParen() {
   return false;
 }
 
+std::optional<std::string_view> Lexer::takeKeyword() {
+  if (curr) {
+    return std::nullopt;
```
Comment (kripken) [inline src/parser/lexer.cpp:1074]: Can you explain this part? It reads as "if there is a current token, ignore it and return nullopt". Why is that correct?

Comment (tlively) [reply src/parser/lexer.cpp:1074]: This goes away at the end of the patch sequence, but it is required for correctness at these intermediate steps. Let's say the lexer reaches an integer and eagerly lexes it into a token stored in `curr`. That integer may be followed by a keyword, but `takeKeyword()` should fail until `takeU32()` (or similar) has been called first. Without this check, `takeKeyword()` would succeed even though the preceeding integer was never consumed.

Comment (kripken) [reply src/parser/lexer.cpp:1074]: I see, thanks. Which is the PR where I can see the final step where it goes away, to get an idea for the direction?

Comment (tlively) [reply src/parser/lexer.cpp:1074]: #6544

**Comment by tlively** [comment]
Looks like this causes the spec tests to hang forever for some reason. Will investigate.

**Comment by tlively** [comment]
### Merge activity

* **Apr 25, 11:47 PM EDT**: @tlively started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/WebAssembly/binaryen/6541).

**Pull request merged**